providers/Spotify/mod.ts

```diff
@@ -0,0 +1,350 @@
+	readonly name = 'Spotify';
+
+	readonly supportedUrls = new URLPattern({
+		hostname: 'open.spotify.com',
+		pathname: '{/intl-:region}?/:type(artist|album)/:id',
+	});
+
+	readonly features: FeatureQualityMap = {
+		'cover size': 640,
+		'duration precision': DurationPrecision.MS,
+		'GTIN lookup': FeatureQuality.GOOD,
+		'MBID resolving': FeatureQuality.GOOD,
+		'release label': FeatureQuality.PRESENT,
+	};
+
+	readonly entityTypeMap = {
+		artist: 'artist',
+		release: 'album',
+	};
+
+	readonly availableRegions = new Set(availableRegions);
+
+	readonly releaseLookup = SpotifyReleaseLookup;
+
+	readonly launchDate: PartialDate = {
+		year: 2008,
+		month: 10,
+	};
+
+	readonly apiBaseUrl = 'https://api.spotify.com/v1/';
+
+	constructUrl(entity: EntityId): URL {
+		return new URL([entity.type, entity.id].join('/'), 'https://open.spotify.com');
+	}
+
+	async query<Data>(apiUrl: URL, maxTimestamp?: number): Promise<CacheEntry<Data>> {
+		try {
+			const accessToken = await this.cachedAccessToken(this.requestAccessToken);
+			const cacheEntry = await this.fetchJSON<Data>(apiUrl, {
+				policy: { maxTimestamp },
+				requestInit: {
+					headers: {
+						'Authorization': `Bearer ${accessToken}`,
+					},
+				},
+			});
+			const apiError = cacheEntry.content as ApiError;
+			if (apiError.error) {
+				throw new SpotifyResponseError(apiError, apiUrl);
+			}
+			return cacheEntry;
+		} catch (error) {
+			// Clone the response so the body of the original response can be
+			// consumed later if the error gets re-thrown.
+			const apiError = await error.response?.clone().json() as ApiError;
+			if (apiError?.error) {
+				throw new SpotifyResponseError(apiError, apiUrl);
+			} else {
+				throw error;
+			}
+		}
+	}
+
+	private async requestAccessToken(): Promise<ApiAccessToken> {
+		// See https://developer.spotify.com/documentation/web-api/tutorials/client-credentials-flow
+		const url = new URL('https://accounts.spotify.com/api/token');
+		const auth = encodeBase64(`${spotifyClientId}:${spotifyClientSecret}`);
+		const body = new URLSearchParams();
+		body.append('grant_type', 'client_credentials');
+
+		const response = await fetch(url, {
+			method: 'POST',
+			headers: {
+				'Authorization': `Basic ${auth}`,
+				'Content-Type': 'application/x-www-form-urlencoded',
+			},
+			body: body,
+		});
+
+		const content = await response.json();
+		return {
+			accessToken: content?.access_token,
+			validUntilTimestamp: Date.now() + (content.expires_in * 1000),
+		};
+	}
+}
+
+export class SpotifyReleaseLookup extends ReleaseApiLookup<SpotifyProvider, Album> {
+	constructReleaseApiUrl(): URL {
+		let lookupUrl: URL;
+		const query = new URLSearchParams();
+		if (this.lookup.region) {
+			query.set('market', this.lookup.region);
+		}
+		if (this.lookup.method === 'gtin') {
+			lookupUrl = new URL(`search`, this.provider.apiBaseUrl);
+			query.set('type', 'album');
+			query.set('q', `upc:${this.lookup.value}`);
```

Comment:
The `market` should only be set for GTIN lookups, otherwise Harmony will use the region which it extracted from the `intl-xx` part of the Spotify URL for URL/ID lookups:
```suggestion
		if (this.lookup.method === 'gtin') {
			lookupUrl = new URL(`search`, this.provider.apiBaseUrl);
			query.set('type', 'album');
			query.set('q', `upc:${this.lookup.value}`);
			if (this.lookup.region) {
				query.set('market', this.lookup.region);
			}
```
With this change it is also no longer necessary to reset `this.lookup.region` to `undefined` after the initial GTIN search.

P.S. Is the `intl-xx` part actually a country code or a language code?
I made this mistake for the Deezer provider where I only noticed it when iTunes requests for the region `EN` were failing :joy:
For Spotify it is a bit unclear, [MBS-13052](https://tickets.metabrainz.org/browse/MBS-13052) assumes it is a language, but for all cases which I have tried where language code and country code would be different, Spotify always seems to redirect to the URL without the intl part (`en` vs `us` or `gb`, `sv` vs `se`), so I can't tell for sure!?
If it is a language, the URLPattern should not use `:region` but `:language` or something else, but as it stands it is currently not necessary to change something because all intl codes which I have seen so far are valid region codes.

Reply:
After doing a few more tests (documented in the comments of the MBS ticket) and lots of indications making me think the opposite, I am now confident this is a language code (`intl-ar` is for Arabic and not for Argentina).

Reply:
See 72bbf236b0407df946ba1ae301f02bdfa0f02a72

Regarding the code: I honestly don't know. I always interpreted it as the market code. But actually it is hard to test. It nearly always redirects to as you say. The redirection seems to be controlled by the Accept-Language header, so that's an argument it is a language and not a market. Do you know any other intl-* paths being used?

```
curl -v -H "Accept-Language: fi" https://open.spotify.com/intl-de/album/2Usbb1Ao4tpzDY0IbIQHqB

< HTTP/2 302 
< location: /album/2Usbb1Ao4tpzDY0IbIQHqB
```

```
curl -v -H "Accept-Language: de" https://open.spotify.com/album/2Usbb1Ao4tpzDY0IbIQHqB

< HTTP/2 302 
< location: /intl-de/album/2Usbb1Ao4tpzDY0IbIQHqB
```

For `curl -v -H "Accept-Language: de" https://open.spotify.com/intl-de/album/2Usbb1Ao4tpzDY0IbIQHqB` there is the  no redirect.

I tested the above with a server located in Finland, just to be sure it behaves the same independent of German location.

I changed it to be `:language` for now. As you say it shouldn't matter, but at least it avoids any issue if this being the region turns out to be wrong.